File: rpsls_j17/src/main/java/nl/bentels/rpsls/PlayerRole.java
package nl.bentels.rpsls;

public sealed interface PlayerRole permits
    Scissors, Paper, Rock, Lizard, Spock {

    int getOrder();

    default MatchResult match(PlayerRole other) {
        final int diff =
                0 <= getOrder() - other.getOrder() ?
                        getOrder() - other.getOrder() :
                        Math.abs(getOrder() - other.getOrder() -1);
        return switch (diff) {
          case 0 -> MatchResult.DRAW;
            default -> switch (diff % 2) {
                case 0 -> MatchResult.WIN;
                case 1 -> MatchResult.LOSE;
                default -> throw new IllegalStateException("Java cannot do maths");
            };
        };
    }
}

final class Scissors implements PlayerRole {

    @Override
    public int getOrder() {
        return 0;
    }
}

final class Paper implements PlayerRole {

    @Override
    public int getOrder() {
        return 1;
    }
}

final class Rock implements PlayerRole {

    @Override
    public int getOrder() {
        return 2;
    }
}

final class Lizard implements PlayerRole {

    @Override
    public int getOrder() {
        return 3;
    }
}

final class Spock implements PlayerRole {

    @Override
    public int getOrder() {
        return 4;
    }
}

File: rpsls_grdl/app/src/main/kotlin/nl/bentels/rpsls/PlayerRole.kt
import kotlin.math.abs
import kotlin.random.Random

enum class PlayerRole(val rank: Int) {
    SCISSORS(0),
    PAPER(1),
    ROCK(2),
    LIZARD(3),
    SPOCK(4);

    fun compareWith(other: PlayerRole): RoleComparisonResult {
        var compareWith: RoleComparisonResult = RoleComparisonResult.DRAW
        if (rank != other.rank) {
            val rankDiff = when (0 < other.rank - rank) {
                true -> other.rank - rank
                false -> abs(other.rank - rank -1)
            }
            compareWith = when (rankDiff % 2) {
                0 -> RoleComparisonResult.LOSE
                1 -> RoleComparisonResult.WIN
                else -> throw IllegalStateException("Kotlin knows nothing of math")
            }
        }
        return compareWith
    }

    companion object Factory {
        private val _random: Random = Random.Default

        internal fun byIndex(i : Int) : PlayerRole =
            when (i) {
                0 -> SCISSORS
                1 -> PAPER
                2 -> ROCK
                3 -> LIZARD
                4 -> SPOCK
                else -> throw NoSuchElementException()
            }

        fun random() = byIndex(_random.nextInt(until = 5))
    }
}